### DISCUSSION TITLE
LoRA Rank vs Learning Rate (In progress...)

### DISCUSSION THREAD

**Discussion by dorpxam** [AkaneTendo25/musubi-tuner#24] [General]
# LoRA Rank vs Learning Rate (In progress...)

## Benchmarking

The common problematic with model training is to find the right parameters to use in order to adjust the ratio quality/training time. LTX-2 is a new model and probably have is own reaction to the common training parameters.

Here, the idea is to attempt to detect the training 'reaction' on two majors parameters :

- LoRA rank
- Learning rate

For simplicity, I will start with a simple 6-photos dataset for a basic Character-LoRA. The optimizer will be the most common `AdamW` and the benchmark will be related to the LoRA rank:

| Rank | Learning rates | Info |
| - | - | - |
| `8` | `2e-4` / `1e-4` / `5e-5` | |
| `16` | `2e-4` / `1e-4` / `5e-5` | |
| `32` | `2e-4` / `1e-4` / `5e-5` | |
| `64` | `2e-4` / `1e-4` / `5e-5` | |
| `128` | `2e-4` / `1e-4` / `5e-5` | if possible, no OOM |

Why these learning rates? Simply because there's the common ones and are half of each other:

| Scientific Notation | Classic Notation | Speed | Info |
| - | - | - | - |
| `2e-4` | `0.00020` | faster | |
| `1e-4` | `0.00010` | medium | half of `2e-4` |
| `5e-5` | `0.00005` | slower | half of `1e-4` |

I hope this will be suffisent to compare the speed of convergence (or not) and the quality difference between the result. It's a hard work, especially to produce a efficient comparison on the produced files. I will probably adjust my methodology if needed.

## Dataset

For a first approach, a basic 6-photos dataset (image-based training) seem perfect to validate the benchmark methology. Maybe in a second approach, a benchmark on video-based / audio-video-based / audio-based methods will be considered.

```toml
[general]
batch_size = 1
enable_bucket = true
bucket_no_upscale = true

[[datasets]]
resolution = [1440, 1792]
image_jsonl_file = "E:/Training/ltx2-rank/dataset/dataset.jsonl"
cache_directory = "E:/Training/ltx2-rank/dataset/cache"
num_repeats = 16
```

Here, the idea is not to have a perfect dataset and result, just enough for doing comparison. So I use few photos of a well-know supermodel.

<details>
<summary>Details</summary>
<img width="1280" height="1062" alt="dataset" src="https://github.com/user-attachments/assets/cf75685c-6ed7-47ef-aaf8-a07c02a532ee" />
</details>
<a>&nbsp;</a>
Very simple, but just enough to capture the essential of the model and doing comparison.

## Parameters

I don't talk here about the basic configuration, I will probably adjust the `--blocks_to_swap` for the different LoRA rank, but the commmon parameters will stay the same. The I use the classic `constant_with_warmup` scheduler with a warmup at 10% of the total number of steps.

```python
--flash_attn
--fp8_base
--fp8_scaled
--blocks_to_swap 10
--gradient_checkpointing
--optimizer_type AdamW
--learning_rate 1e-4
--lr_scheduler constant_with_warmup
--lr_warmup_steps 230
--network_module networks.lora_ltx2
--network_dim 16
--network_alpha 16
--timestep_sampling shifted_logit_normal
--max_train_epochs 24
--save_every_n_epochs 1
```

I start with ~2000 steps (2304 to be precise), but if needed, I will push the training to more steps/epochs. I need to test the current dataset.

**RANK**

`--network_dim`

Type: `integer`
Values: `8 | 16 | 32 | 64 | 128` 
Default: `32`

The rank of the LoRA adaptation matrices. Controls the capacity (expressiveness) of the adapter.

| Rank  | Best For |
| - | - |
| 8-16 | Simple style transfer, subtle adjustments |
| 32 | General purpose, recommended starting point |
| 64-128 | Complex concepts if lower ranks don't give good results |

`--network_alpha`

For this benchmark `--network_alpha` same as `--network_dim`

**LEARNING RATE**

`--learning_rate`

Type: `float` 
Range: `1e-6 to 1.0` 
Default: `2e-4`

Initial learning rate for the optimizer. The default value works well for most use cases.

**NUMBER OF STEPS**

`--max_train_steps`

Type: `integer`
Range: `100 to 20,000` 
Default: `2000`

Total number of training optimization steps.

| Steps	| Use Case |
| - | - |
| 1000-2000	| Style transfer with 10-20 videos |
| 2000-6000	| Learning specific subjects/characters |
| 6000+	| Large datasets only (risk of overfitting otherwise) |

## Steps vs Epochs

I prefer using `--max_train_epochs` vs `--max_train_steps`. The number of steps is related to the `num train items` (aka. `num media` x `num repeats`)

In this benchmark, image-based, there's only `6 photos` (resolution: `1440x1792`) repeated `16 times` in one epoch. So `1 epoch` is `96 items`, I can simple approximate the number of steps :

`mun_train_epochs = 2000 steps / 96 items`, so `~ 20.8333...` rounded to a upper power of 2 (love that) : `24`

I can fix my `--max_train_epochs` to `24`

The true number of steps become `24 x 96 = 2304`, in fact if you want to approximate `~2000` steps, so `21 x 96 = 2016` was more close. 

Slightly upper the default `--max_train_steps 2000`, but fit with the dataset items while settings `--max_train_steps` to `2000` do no fit regular epoch, remember : `2000 / 96 = 20.8333...`
